apps/els_lsp/src/els_refactorerl_diagnostics.erl

```diff
@@ -0,0 +1,131 @@
+                                          ([els_diagnostics:diagnostic()]).
+run_query(Module, DiagnosticId) ->
+  {_, Message, _, _} = DiagnosticId,
+  ReferlResult = els_refactorerl_utils:query(make_query(DiagnosticId, Module)),
+  Pois = els_refactorerl_utils:convert_to_poi(ReferlResult),
+  [make_diagnostic(Poi, Message) || Poi <- Pois].
+
+-spec is_default() -> boolean().
+is_default() ->
+  true.
+
+-spec run(uri()) -> [els_diagnostics:diagnostic()].
+run(Uri) ->
+  run(Uri, 0).
+
+
+-spec run(uri(), number()) -> [els_diagnostics:diagnostic()].
+run(Uri, RecursionDepth) when RecursionDepth < ?MAX_RECURSION_DEPTH ->
+  case filename:extension(Uri) of
+    <<".erl">> ->
+      case els_refactorerl_utils:referl_node() of
+        disabled ->
+          [];
+        _ ->
+          case add(Uri) of
+            busy ->
+              timer:sleep(1000),
+              run(Uri, RecursionDepth + 1);
+            disabled ->
+              [];
+            _ ->
+              FileName = filename:basename(binary_to_list(els_uri:path(Uri))),
+              Module = list_to_atom(filename:rootname(FileName)),
+              Diagnostics = refactorerl_diagnostics(),
+              lists:concat([run_query(Module, DiagId) || DiagId <- Diagnostics])
+          end
+      end;
+    _ ->
+      []
+  end;
+
+run(_, RecursionDepth) when RecursionDepth >= ?MAX_RECURSION_DEPTH ->
+  Msg = "Cannot add module to RefactorErl!",
+  els_refactorerl_utils:notification(Msg, ?MESSAGE_TYPE_ERROR),
+  [].
+
+-spec source() -> binary().
+source() ->
+  <<"RefactorErl Diagnostics">>.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+%%@doc
+%% Creates a diagnostic form the Poi data and a Message
+%% The severity is only warning.
+-spec make_diagnostic(poi(), [char()]) -> els_diagnostics:diagnostic().
+make_diagnostic(#{ data := PoiData, range := PoiRange}, DiagMessage) ->
+    Range = els_protocol:range(PoiRange),
+    Message = list_to_binary(DiagMessage ++ " " ++ PoiData),
+    Severity = ?DIAGNOSTIC_WARNING,
+    Source = source(),
+    els_diagnostics:make_diagnostic(Range, Message, Severity, Source).
+
+  %%@doc
+  %% Returns the available diagnostics of RefactorErl.
+-spec refactorerl_diagnostics() -> [refactorerl_diagnostic_id()].
+refactorerl_diagnostics() -> % TODO: Make it configureable
+  [ {unused_calls, "Security Issue",  "mods[name=", "].funs.unsecure_calls"}
+  ,   {unsecure_macros
+      , "Unused Macros:"
+      , "mods[name="
+      , "].macros[not .references]" }
+  ].
+
+  %%@doc
+  %% Adds a module to the RefactorErl node.
+-spec add(any()) -> atom(). %TODO: Add .hrl files
+add(Uri) ->
+  Param = [binary_to_list(els_uri:path(Uri))],
+  case els_refactorerl_utils:referl_node() of
+    {ok, Node} ->
+      rpc:call(Node, ri, add, Param, els_refactorerl_utils:maxtimeout());
```

Comment:
What does the `ri:add` function returns? Do we want to handle errors?

Reply:
This is changed a bit, as the request to the RefactorErl node was not so stable. I will describe this in a comment, when I merge to this PR.